enhancements/monitoring/alert-overrides.md

```diff
@@ -0,0 +1,280 @@
+alerting rules.
+
+Users currently have the ability to add additional rules from a technical
+standpoint -- by simply creating new `PrometheusRule` objects -- but this is
+explicitly unsupported at the moment.  The reason being that these resources are
+handled directly by the prometheus-operator, making it difficult to perform
+OpenShift specific defaulting and validation.  A concrete example being that
+there is no way to prevent users from creating new recording rules that may
+overwhelm the platform Prometheus instance.
+
+We've also encountered bugs and unexpected interactions that cause alerts to
+fire erroneously.  A supported method of patching platform alerting rules
+provides a more flexible method of applying temporary fixes in such cases.
+
+The goal of this proposal is to enable the addition of custom alerting rules,
+and limited ability to override existing platform alerting rules via a resource
+under the direct control of the cluster-monitoring-operator.
+
+### Goals
+
+- Give users a single, supported, and documented place to add additional rules,
+  in a way that allows the cluster-monitoring-operator to provide its own
+  defaulting, validation, etc.
+- Allow users to apply overrides to alerting rules shipped by the platform.
+- Allow users to drop unwanted platform alerts.
+- User-supplied overrides are merged with platform rules -- updates to fields
+  not modified by users continue to get updates.
+- Resources shipped by operators remain unmodified.  No action needed from
+  OpenShift developers.
+
+### Non-Goals
+
+- Only alerting rules are available for modification.  Not recording rules or
+  other monitoring configuration.
+
+## Proposal
+
+The cluster-monitoring-operator will be extended with a new controller that
+watches a user-supplied `ConfigMap` with a single key having the following
+structure:
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: alert-overrides
+  namespace: openshift-monitoring
+data:
+  overrides.yaml: |
+    overrides: # List of platform rule overrides.
+    - key:
+        name: "Watchdog" # Alerts are identified by name and severity tuple.
+        severity: "none"
+      action: "patch"
+      labels:
+        patched: true
+    - key:
+        name: "AlertmanagerReceiversNotConfigured"
+        severity: "Warning"
+      action: "drop"
+
+    rules: # List of additional alerting rules.
+    - alert: "CustomAlert"
+      expr: "vector(1)"
+```
+
+The controller maintains an index of alerting rule name-and-severity pairs to
+`PrometheusRule` objects, ignoring any in non-platform namespaces, i.e. any
+namespace without the `openshift.io/cluster-monitoring=true` label.  When the
+configuration changes, the `PrometheusRule` containing the targeted alert is
+found, and the alert extracted.
+
+For each override a new alerting rule is generated with the overrides applied on
+top of the existing rule, and this new rule is added to the "alert-overrides"
```

Comment:
I am having a hard time grok'ing this. Could you please clarify?

- CM exists with identifiers and actions
- Existing PrometheusRules are identified using the identifiers found in the CM
- Actions are applied (Patch,Drop?) to these found PrometheusRules
- Replacement rules are added to a PrometheusRules object called "alert-overrides"
- label is add to each new replacement rule
?

Reply:
Yeah, that's basically it. For each entry in the `overrides` section of the `ConfigMap`, find the matching rule in the existing `PrometheusRule` objects, then apply the action -- that is either `drop` or `patch`.

If the action is `patch`, apply the user's changes on top of the existing rule, and add the result to the "alert-overrides" `PrometheusRule`. These always get a static label identifying them as a generated override rule. That label is what allows us to drop the original rule in the relabel config, because we can match only rules that *don't* have the label.

So, the last step is to then generate the `alert_relabel_configs` which drops the original version of any overridden alerts and any alerts the user specified the `drop` action for.

Reply:
I've added a bit trying to summarize this better in the proposal section.